rust/user-info-fetcher/src/backend/active_directory.rs

```diff
@@ -0,0 +1,394 @@
+            Error::FindUserLdap { .. } => StatusCode::SERVICE_UNAVAILABLE,
+            Error::FindUserGroupsLdap { .. } => StatusCode::SERVICE_UNAVAILABLE,
+            Error::ParseIdByClient { .. } => StatusCode::BAD_REQUEST,
+            Error::ParseIdByLdap { .. } => StatusCode::INTERNAL_SERVER_ERROR,
+            Error::UserNotFound { .. } => StatusCode::NOT_FOUND,
+            Error::InvalidPrimaryGroupRelativeId { .. } => StatusCode::INTERNAL_SERVER_ERROR,
+            Error::UserSidHasNoSubauthorities { .. } => StatusCode::INTERNAL_SERVER_ERROR,
+            Error::ParseUserSid { .. } => StatusCode::INTERNAL_SERVER_ERROR,
+        }
+    }
+}
+
+// Matching rules defined at https://learn.microsoft.com/en-us/windows/win32/adsi/search-filter-syntax#operators
+/// Makes DN filters apply recursively to group membership
+const LDAP_MATCHING_RULE_IN_CHAIN: &str = ":1.2.840.113556.1.4.1941:";
+
+const LDAP_FIELD_OBJECT_ID: &str = "objectGUID";
+const LDAP_FIELD_OBJECT_SECURITY_ID: &str = "objectSid";
+const LDAP_FIELD_OBJECT_DISTINGUISHED_NAME: &str = "dn";
+const LDAP_FIELD_USER_NAME: &str = "userPrincipalName";
+const LDAP_FIELD_USER_PRIMARY_GROUP_RID: &str = "primaryGroupID";
+const LDAP_FIELD_GROUP_MEMBER: &str = "member";
+
+pub(crate) async fn get_user_info(
+    request: &UserInfoRequest,
+    ldap_server: &str,
+    tls: &TlsClientDetails,
+    base_distinguished_name: &str,
+    custom_attribute_mappings: &BTreeMap<String, String>,
+) -> Result<UserInfo, Error> {
+    let ldap_tls = utils::tls::configure_native_tls(tls)
+        .await
+        .context(ConfigureTlsSnafu)?;
+    let (ldap_conn, mut ldap) = LdapConnAsync::with_settings(
+        LdapConnSettings::new().set_connector(ldap_tls),
+        &format!(
+            "{protocol}://{ldap_server}",
+            protocol = if tls.uses_tls() { "ldaps" } else { "ldap" }
+        ),
+    )
+    .await
+    .context(ConnectLdapSnafu)?;
+    ldap3::drive!(ldap_conn);
+    ldap.sasl_gssapi_bind(ldap_server)
+        .await
+        .context(RequestLdapSnafu)?
+        .success()
+        .context(BindLdapSnafu)?;
+    let user_filter = match request {
+        UserInfoRequest::UserInfoRequestById(id) => {
+            format!(
+                "{LDAP_FIELD_OBJECT_ID}={}",
+                ldap_escape_bytes(
+                    &Uuid::from_str(&id.id)
+                        .context(ParseIdByClientSnafu)?
+                        .to_bytes_le()
+                )
+            )
+        }
+        UserInfoRequest::UserInfoRequestByName(username) => {
+            format!("{LDAP_FIELD_USER_NAME}={}", ldap_escape(&username.username))
+        }
+    };
+    let requested_user_attrs = [
+        LDAP_FIELD_OBJECT_SECURITY_ID,
+        LDAP_FIELD_OBJECT_ID,
+        LDAP_FIELD_USER_NAME,
+        LDAP_FIELD_USER_PRIMARY_GROUP_RID,
+    ]
+    .into_iter()
+    .chain(custom_attribute_mappings.values().map(String::as_str))
+    .collect::<Vec<&str>>();
+    let user = ldap
```

Comment:
Would it make sense to add a debug/trace statement here with what attributes where requested and what was returned? 
I manually added these while reviewing and debugging why it didn't work.

Reply:
3fc772a478228e120a65ef5058beda8385151eaa